In [7]:
import pandas as pd
import pandas_datareader as web
import seaborn as sns
from datetime import datetime

In [5]:
f = web.DataReader('USD000UTSTOM', 'moex', start='2017-07-01', end='2017-07-31')
f.head()

BOARDID    CLOSE     HIGH     LOW NUMTRADES   OPEN         SECID  \
TRADEDATE                                                                     
2017-07-03    CETS  59.2650  59.4825  58.790     29108  58.95  USD000UTSTOM   
2017-07-03    CNGD  59.3600  59.4250  58.840        24  58.98  USD000UTSTOM   
2017-07-04    CETS  59.4125  59.4575  59.135     21053  59.30  USD000UTSTOM   
2017-07-04    CNGD  59.3575  59.3600  58.930        37  59.36  USD000UTSTOM   
2017-07-05    CETS  59.9825  60.2600  59.300     50108  59.30  USD000UTSTOM   

             SHORTNAME        VOLRUR  WAPRICE  
TRADEDATE                                      
2017-07-03  USDRUB_TOM  1.472424e+11  59.1903  
2017-07-03  USDRUB_TOM  1.864785e+09      NaN  
2017-07-04  USDRUB_TOM  1.090265e+11  59.2700  
2017-07-04  USDRUB_TOM  1.046416e+09      NaN  
2017-07-05  USDRUB_TOM  2.874226e+11  59.9234

In [12]:
sber=web.DataReader('SBER', 'moex', start='2017-01-01', end='2017-12-31')
sber.head(5)

ADMITTEDQUOTE  ADMITTEDVALUE  BEICLOSE  BID BOARDID BOARDNAME  \
TRADEDATE                                                                   
2017-01-03            NaN            NaN       NaN  NaN    RPMA       NaN   
2017-01-04            NaN            NaN       NaN  NaN    RPMA       NaN   
2017-01-05            NaN            NaN       NaN  NaN    RPMA       NaN   
2017-01-06            NaN            NaN       NaN  NaN    RPMA       NaN   
2017-01-09            NaN            NaN       NaN  NaN    RPMA       NaN   

            BUYBACKDATE  CBRCLOSE   CLOSE CLOSEAUCTIONPRICE  ... TRENDCLSPR  \
TRADEDATE                                                    ...              
2017-01-03          NaN       NaN  169.74               NaN  ...        NaN   
2017-01-04          NaN       NaN  169.74               NaN  ...        NaN   
2017-01-05          NaN       NaN  169.74               NaN  ...        NaN   
2017-01-06          NaN       NaN  169.74               NaN  ...        NaN   
2017-01-09          NaN       NaN  171.85               NaN  ...        NaN   

           TRENDWAP  TRENDWAPPR  TYPE         VALUE VOLUME  WAPRICE WAVAL  \
TRADEDATE                                                                   
2017-01-03      NaN         NaN   NaN  7.982129e+08    NaN      NaN   NaN   
2017-01-04      NaN         NaN   NaN  8.429459e+08    NaN      NaN   NaN   
2017-01-05      NaN         NaN   NaN  8.436439e+08    NaN      NaN   NaN   
2017-01-06      NaN         NaN   NaN  8.423766e+08    NaN      NaN   NaN   
2017-01-09      NaN         NaN   NaN  8.618491e+08    NaN      NaN   NaN   

            YIELDLASTCOUPON  YIELDTOOFFER  
TRADEDATE                                  
2017-01-03              NaN           NaN  
2017-01-04              NaN           NaN  
2017-01-05              NaN           NaN  
2017-01-06              NaN           NaN  
2017-01-09              NaN           NaN  

[5 rows x 65 columns]

In [19]:
sber_drop = sber.dropna(axis='columns', how = 'all')
display(sber_drop.BOARDID.unique()) 
sber_drop

array(['RPMA', 'SOTC', 'EQRP', 'PSRP', 'EQWP', 'EQRD', 'PSRD', 'MXBD',
       'PSEQ', 'PTEQ', 'EQDP', 'SMAL', 'TQBR'], dtype=object)

ADMITTEDQUOTE  ADMITTEDVALUE  BID BOARDID   CLOSE CURRENCYID  \
TRADEDATE                                                                  
2017-01-03            NaN            NaN  NaN    RPMA  169.74        NaN   
2017-01-04            NaN            NaN  NaN    RPMA  169.74        NaN   
2017-01-05            NaN            NaN  NaN    RPMA  169.74        NaN   
2017-01-06            NaN            NaN  NaN    RPMA  169.74        NaN   
2017-01-09            NaN            NaN  NaN    RPMA  171.85        NaN   
...                   ...            ...  ...     ...     ...        ...   
2017-12-28            NaN   0.000000e+00  NaN    SMAL  223.27        NaN   
2017-12-28          224.2   5.412390e+09  NaN    TQBR  224.20        NaN   
2017-12-29            NaN   0.000000e+00  NaN    EQDP     NaN        NaN   
2017-12-29            NaN   0.000000e+00  NaN    SMAL  225.61        NaN   
2017-12-29          225.2   4.258428e+09  NaN    TQBR  225.20        NaN   

            DECIMALS FACEUNIT  FACEVALUE     HIGH  ... OFFER    OPEN  \
TRADEDATE                                          ...                 
2017-01-03       NaN      NaN        NaN  186.714  ...   NaN  173.25   
2017-01-04       NaN      NaN        NaN  186.714  ...   NaN  180.12   
2017-01-05       NaN      NaN        NaN  186.714  ...   NaN  175.95   
2017-01-06       NaN      NaN        NaN  186.714  ...   NaN  171.90   
2017-01-09       NaN      NaN        NaN  189.035  ...   NaN  170.69   
...              ...      ...        ...      ...  ...   ...     ...   
2017-12-28       NaN      NaN        NaN  224.590  ...   NaN  224.26   
2017-12-28       NaN      NaN        NaN  227.170  ...   NaN  226.90   
2017-12-29       NaN      NaN        NaN      NaN  ...   NaN     NaN   
2017-12-29       NaN      NaN        NaN  225.960  ...   NaN  225.01   
2017-12-29       NaN      NaN        NaN  226.200  ...   NaN  224.12   

            REGNUMBER  SECID  SHORTNAME  TRADINGSESSION  TYPE         VALUE  \
TRADEDATE                                                                     
2017-01-03        NaN   SBER   Сбербанк             3.0   NaN  7.982129e+08   
2017-01-04        NaN   SBER   Сбербанк             3.0   NaN  8.429459e+08   
2017-01-05        NaN   SBER   Сбербанк             3.0   NaN  8.436439e+08   
2017-01-06        NaN   SBER   Сбербанк             3.0   NaN  8.423766e+08   
2017-01-09        NaN   SBER   Сбербанк             3.0   NaN  8.618491e+08   
...               ...    ...        ...             ...   ...           ...   
2017-12-28        NaN   SBER   Сбербанк             3.0   NaN  1.722529e+04   
2017-12-28        NaN   SBER   Сбербанк             3.0   NaN  5.412390e+09   
2017-12-29        NaN   SBER   Сбербанк             3.0   NaN  0.000000e+00   
2017-12-29        NaN   SBER   Сбербанк             3.0   NaN  7.216460e+03   
2017-12-29        NaN   SBER   Сбербанк             3.0   NaN  4.258428e+09   

                VOLUME  WAPRICE  
TRADEDATE                        
2017-01-03         NaN      NaN  
2017-01-04         NaN      NaN  
2017-01-05         NaN      NaN  
2017-01-06         NaN      NaN  
2017-01-09         NaN      NaN  
...                ...      ...  
2017-12-28        77.0   224.77  
2017-12-28  24078940.0   224.77  
2017-12-29         0.0      NaN  
2017-12-29        32.0   225.08  
2017-12-29  18919860.0   225.08  

[2489 rows x 30 columns]

In [2]:
from finam import (Exporter,
                   Timeframe,
                   Market,
                   FinamExportError,
                   FinamObjectNotFoundError)
exporter = Exporter()

In [3]:
exporter.lookup(code='LKOH')

name  code  market
id                                      
8                   ЛУКОЙЛ  LKOH       1
332                   LKOH  LKOH       3
918                 Лукойл  LKOH       8
17453                 LKOH  LKOH      14
22427   LKOH(RTS-Standard)  LKOH      38
8                   ЛУКОЙЛ  LKOH     200
419710  ПАО \'ЛУКОЙЛ\', ао  LKOH     517

Коды рынков на экспорт (не все, остальные смотрите на finam.ru, но они не качаютcя)    
    BONDS = 2<br>
    COMMODITIES = 24<br>
    CURRENCIES = 45<br>
    ETF = 28<br>
    ETF_MOEX = 515<br>
    FUTURES = 14  # non-expired futures<br>
    FUTURES_ARCHIVE = 17  # expired futures<br>
    FUTURES_USA = 7<br>
    INDEXES = 6<br>
    SHARES = 1<br>
    SPB = 517<br>
    USA = 25<br>

In [4]:
# посмотрим что есть на рынках ММВБ
market = exporter.lookup(market=Market.SHARES)
display(market.head())
print(','.join(market['code']))

name   code  market
id                               
388313   Роллман-п  RLMNP       1
22843   iАвиастКао   UNAC       1
473000  Сахэнер ао   SLEN       1
13      Сургнфгз-п  SNGSP       1
24      СаратЭн-ап  SAREP       1

RLMNP,UNAC,SLEN,SNGSP,SAREP,VLHZ,SARE,SNGS,VTBR,KRKNP,KRKN,DIOD,ODVA,MFGSP,PLZL,MFGS,URKZ,STSBP,STSB,USBN,SMLT,MFON,ISKJ,GEMA,NAUK,PMSB,TRNFP,MOEX,MORI,VJGZP,VJGZ,MRSB,PIKK,EELT,T-RM,MTLRP,URKA,LIFE,AAPL-RM,SBUX-RM,ZVEZ,PRTK,PAZA,ZILL,TGKA,TANLP,TANL,PMSBP,MTLR,TGKN,GOOG-RM,AMD-RM,PHOR,BIDU-RM,AFKS,INGR,MRKC,MRKP,SFIN,FIVE,IRKT,HIMC,HIMCP,MRKU,WTCM,WTCMP,MRKZ,IGSTP,RZSB,RNFT,RUSP,IGST,ROLO,MGNZ,JNOSP,BA-RM,JNOS,SAGO,IBM-RM,FLOT,PRMB,BMY-RM,MGNT,MAGEP,MAGE,XOM-RM,FEES,SAGOP,MTSS,BIIB-RM,TUZA,RBCM,RGSS,RDRB,RAVN,KHC-RM,NPOF,SVAV,RUSI,QIWI,RUGR,VZRZ,MOBB,YRSB,YRSBP,TNSE,AAL-RM,CSCO-RM,AMZN-RM,TORS,TORSP,V-RM,TASB,TASBP,TWTR-RM,TSLA-RM,TGT-RM,NVTK,TATN,TATNP,NKSH,TTLK,JNJ-RM,NFAZ,RTSBP,RTSB,NNSBP,NNSB,OBUV,OMZZP,KRKO,UBER-RM,KRKOP,UCSS,SIBN,HALS,TUCH,CVX-RM,VTRS-RM,OGKB,UWGN,TRMK,FESH,RTGZ,GTRK,KBSB,MISBP,VRSBP,VRSB,MISB,HPQ-RM,VZRZP,CNTL,ADBE-RM,GM-RM,GRNT,TRFM,GTSS,GLTR,GAZP,GAZC,GAZS,GAZT,GAZAP,GAZA,MSST,GTLC,MSTT,VSYDP,VSYD,TGKBP,TGKB,GE-RM,VGSBP,VGSB,GMKN,TRCN,DZRDP,DZRD,KO-RM,WMT-RM,

In [63]:
# посмотрим что есть на товарных рынках
market = exporter.lookup(market=Market.COMMODITIES)
display(market.head())
print(','.join(market['code']))

name          code  market
id                                         
2131518  Brent(Sep25)  Brent.25.Sep      24
912183   Brent(Sep24)  Brent.24.Sep      24
2017857  Light(Jul25)  Light.25.Jul      24
485788   Brent(May22)  Brent.22.May      24
74454           Сахар            YO      24

Brent.25.Sep,Brent.24.Sep,Light.25.Jul,Brent.22.May,YO,CL,Brent.23.May,Light.24.Dec,RB,Brent.23.Sep,Light.20.Dec,Light.24.Feb,Light.21.Dec,Light.22.Dec,Light.25.Feb,Brent.25.May,Brent.22.Sep,Brent.21.Mar,Light.23.Dec,Brent.21.Sep,Brent.24.Nov,Brent.24.May,Light.23.Jan,Brent.23.Nov,Brent.20.Nov,Brent.25.Oct,Light.23.Jul,Brent.24.Oct,Brent.23.Oct,PB,Brent.21.Oct,Brent.21.Aug,Brent.25.Apr,Light.21.Jul,Light.23.Feb,Brent.23.Apr,Brent.22.Apr,Brent.21.Apr,Light.22.Jul,PA,Brent.24.Apr,SN,PL,Brent.23.Aug,Light.25.May,Brent.22.Nov,AH,Light.24.Jan,Light.24.May,BZ,Light.22.Jan,Brent.22.Aug,Light.23.May,Light.22.May,Light.24.Jul,Light.21.May,Brent.21.Nov,SI,Brent.25.Aug,Brent.24.Aug,Light.21.Jan,Brent.22.Oct,Light.22.Apr,Brent.21.May,HO,Brent.21.Jul,Brent.24.Dec,Light.22.Sep,Brent.23.Dec,Brent.22.Dec,Brent.21.Dec,Brent.20.Dec,Light.21.Sep,CA,Brent.22.Jul,GC,Light.23.Sep,Light.25.Oct,Light.24.Sep,Light.25.Sep,Light.24.Nov,Light.24.Oct,Light.21.Feb,Brent.22.Mar,Light.24.Apr,Light.23.Mar,Brent.24.Mar

In [8]:
# тикер BRENT
market = exporter.lookup(code = "BZ", market=Market.COMMODITIES)
display(market.head())

name code  market
id                       
19473  Brent   BZ      24

In [9]:
start_date = datetime(2017, 1, 1)
end_date = datetime(2017, 12, 31)
sber = exporter.lookup(code='SBER',market=Market.SHARES)

display(sber)

data_daily = exporter.download(
    sber.index[0], 
    market=Market.SHARES,
    start_date=start_date,
    end_date=end_date
)
data_daily.head()


name  code  market
id                        
3   Сбербанк  SBER       1

<DATE>    <TIME>  <OPEN>  <HIGH>   <LOW>  <CLOSE>     <VOL>
0  20170103  00:00:00  173.41  181.68  173.00   180.12  53842180
1  20170104  00:00:00  178.91  179.41  174.23   175.95  41327620
2  20170105  00:00:00  176.32  177.28  171.54   171.90  35076900
3  20170106  00:00:00  172.41  173.75  169.50   170.69  35395940
4  20170109  00:00:00  171.37  172.39  168.05   168.35  35166440

In [13]:
#Timeframe:
# TICKS
# MINUTES1
# MINUTES5
# MINUTES10
# MINUTES15
# MINUTES30
# HOURLY
# DAILY
# WEEKLY
# MONTHLY

data_5min = exporter.download(
    #id_ = sber.index[0],
    id_ = 3,
    market = Market.SHARES,
    start_date = datetime(2020, 1, 1),
    end_date = datetime(2020, 1, 5),
    timeframe=Timeframe.MINUTES5 
)
data_5min.head(10)

<DATE>    <TIME>  <OPEN>  <HIGH>   <LOW>  <CLOSE>    <VOL>
0  20200103  10:00:00  255.97  256.97  255.85   256.77  1161410
1  20200103  10:05:00  256.77  256.90  256.30   256.70   910890
2  20200103  10:10:00  256.68  256.75  256.03   256.12   440290
3  20200103  10:15:00  256.19  256.42  255.83   255.89   472330
4  20200103  10:20:00  255.89  256.31  255.88   256.13   219100
5  20200103  10:25:00  256.19  256.31  256.08   256.29   159720
6  20200103  10:30:00  256.20  256.32  256.09   256.22   132250
7  20200103  10:35:00  256.21  256.21  256.01   256.15   109490
8  20200103  10:40:00  256.15  256.46  256.08   256.43    99440
9  20200103  10:45:00  256.41  256.47  256.32   256.38    84220

Поработаем с дневными данными

In [87]:
data_daily.head(5)

<DATE>    <TIME>  <OPEN>  <HIGH>   <LOW>  <CLOSE>     <VOL>
0  20170103  00:00:00  173.41  181.68  173.00   180.12  53842180
1  20170104  00:00:00  178.91  179.41  174.23   175.95  41327620
2  20170105  00:00:00  176.32  177.28  171.54   171.90  35076900
3  20170106  00:00:00  172.41  173.75  169.50   170.69  35395940
4  20170109  00:00:00  171.37  172.39  168.05   168.35  35166440

In [94]:
data_daily.columns = ['Date','Time','Open', 'High','Low','Close','Volume']
data_daily.index = pd.to_datetime(data_daily.Date,format='%Y%m%d')
data_daily.drop(columns=['Date','Time'],inplace =True)
data_daily.head(5)


Open    High     Low   Close    Volume
Date                                                
2017-01-03  173.41  181.68  173.00  180.12  53842180
2017-01-04  178.91  179.41  174.23  175.95  41327620
2017-01-05  176.32  177.28  171.54  171.90  35076900
2017-01-06  172.41  173.75  169.50  170.69  35395940
2017-01-09  171.37  172.39  168.05  168.35  35166440

In [96]:
data_volume_month=data_daily.Volume.resample('m').sum()
data_volume_month

Date
2017-01-31     988723570
2017-02-28     816284690
2017-03-31     978856860
2017-04-30     963970030
2017-05-31     824194850
2017-06-30    1247585930
2017-07-31    1055435570
2017-08-31    1063813850
2017-09-30     942566460
2017-10-31     744449580
2017-11-30    1252529590
2017-12-31     682568440
Freq: M, Name: Volume, dtype: int64

In [97]:
open_m=data_daily.Open.resample('M').first().values
open_m

array([173.41, 172.51, 155.99, 160.5 , 166.  , 156.4 , 146.69, 165.53,
       183.85, 192.71, 194.5 , 224.53])

In [98]:
close_m=data_daily.Close.resample('M').last().values
low_m=data_daily.Low.resample('M').min().values
high_m=data_daily.High.resample('M').max().values

In [99]:
month=data_daily.index.strftime('%b').unique()
month

Index(['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct',
       'Nov', 'Dec'],
      dtype='object', name='Date')

In [100]:
data_daily.to_period('M').index.unique()

PeriodIndex(['2017-01', '2017-02', '2017-03', '2017-04', '2017-05', '2017-06',
             '2017-07', '2017-08', '2017-09', '2017-10', '2017-11', '2017-12'],
            dtype='period[M]', name='Date', freq='M')

In [101]:
moth_a = pd.DataFrame(
    {'month':month,'open':open_m,'close':close_m,'low':low_m,'high':high_m},
    index=data_daily.to_period('M').index.unique())
moth_a.head(3)

month    open  close     low    high
Date                                        
2017-01   Jan  173.41  172.2  163.34  181.68
2017-02   Feb  172.51  156.0  155.70  174.98
2017-03   Mar  155.99  159.8  155.23  165.99

In [103]:
sb_m = data_daily.resample('m').agg({'Open':'first','Low':'min','High':'max','Close':'last'})
sb_m

Open     Low    High   Close
Date                                      
2017-01-31  173.41  163.34  181.68  172.20
2017-02-28  172.51  155.70  174.98  156.00
2017-03-31  155.99  155.23  165.99  159.80
2017-04-30  160.50  147.62  168.45  165.20
2017-05-31  166.00  155.03  172.39  155.93
2017-06-30  156.40  136.20  158.89  145.59
2017-07-31  146.69  146.22  167.35  164.53
2017-08-31  165.53  165.30  186.33  183.51
2017-09-30  183.85  180.93  195.00  192.33
2017-10-31  192.71  189.06  198.52  193.80
2017-11-30  194.50  192.38  233.95  224.35
2017-12-31  224.53  219.23  230.50  225.20

In [105]:
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly.graph_objs as go
init_notebook_mode(connected=True)

In [106]:
trace = go.Ohlc(x=data_daily.index,
                open=data_daily['Open'],
                high=data_daily['High'],
                low=data_daily['Low'],
                close=data_daily['Close'])

data = [trace]
iplot(data)

In [116]:
trace2 = go.Candlestick(
                x=sb_m.index,# Проблема месяц, есть ли решение на середину месяца
                open=sb_m['Open'],
                high=sb_m['High'],
                low=sb_m['Low'],
                close=sb_m['Close'])

# убираем нижний бегунок
layout = go.Layout(
    xaxis = dict(
        rangeslider = dict(
            visible = False
        )
    )
)

data2 = [trace2]
fig = go.Figure(data=data2,layout=layout)
iplot(fig)

In [114]:
month=sb_m.index.strftime('%b-%Y')

trace2 = go.Candlestick(
                x=month, # Внимание на месяц
                open=sb_m['Open'],
                high=sb_m['High'],
                low=sb_m['Low'],
                close=sb_m['Close'])

# убираем нижний бегунок
layout = go.Layout(
    xaxis = dict(
        rangeslider = dict(
            visible = False
        )
    )
)

data2 = [trace2]
fig = go.Figure(data=data2,layout=layout)
iplot(fig)